# FunGAP: Fungal Genome Annotation Pipeline

FunGAP performs gene prediction on given genome assembly and RNA-seq reads. See **INSTALL.md** and **USAGE.md** for installation and usage, or you can go wiki tab for the same.

* [Download FunGAP](#download)
* [BLAST+ installation](#blast)
* [Trinity installation](#trinity)
  * [Maker2 installation](#maker)

## FunGAP INPUT & OUTPUT

FunGAP inputs:
```
--output_dir                      | Output directory
--trans_read_files                | Illumina paired-end mRNA reads files (FASTQ)
--project_name                    | Project name without white space
--genome_assembly                 | Genome assembly file (FASTA)
--augustus_species                | Augustus --species argument
--org_id                          | Organism ID for assigning gene IDs
--sister_proteome                 | Protein database (FASTA)
--num_cores                       | Number of CPU cores to be used
```
FunGAP outputs:
```
fungap_output.gff3                | Tab-delimited genomic feature format
fungap_output_prot.faa            | Translated protein sequences
fungap_output_result_summary.html | Summary of FunGAP results
```

## Pipeline description

![](http://compbio.korea.ac.kr/bnmin/fungap/scheme_fungap_ver2.png)

### 1. Step1: Preprocessing of input data
In preprocessing step, FunGAP masks repeat regions in genome assembly (input data 1) and assembles mRNA reads into transcript contigs (input data 2). 

**Repeat masking**

Repeat masking is a crucial step in eukaryotic gene prediction because genomic regions, such as transposon repeats, often make false alignments and interfere with gene prediction (Yandell and Ence, 2012). FunGAP employs a repeat masking procedure embedded in the Maker pipeline (Cantarel, et al., 2008) along with a genome-specific repeat library built by RepeatModeler (http://www.repeatmasker.org/RepeatModeler.html).

**Assembly of mRNA reads**

User-provided mRNA reads are assembled by the Trinity program (Haas, et al., 2013). A BAM-format file for genome-guided assembly is generated by a Hisat2 read aligner (Kim, et al., 2015) and Samtools format converter (SAM file to sorted BAM file) (Li, et al., 2009). An optional parameter (--jaccard_clip) in Trinity is used for fungal transcript assembly because high gene density leads to UTR overlap in the assembly. This option helps avoid fusion of neighbor transcripts. The maximum intron length is set to 2000 bp with the --max-intronlen option in Hisat2.



## Contact

* Project principal investigator: In-Geol Choi at Korea University
* Contact: mbnmbn00@korea.ac.kr (or mbnmbn00@gmail.com)

If you have any problem to install or run, please don't hesistate to contact me. I will help you as much as I can and that helps build more robust program.



![](http://compbio.korea.ac.kr/bnmin/fungap/fungap_logo-01.png)

