#!/usr/bin/python

'''
Run Hisat2 for intron-aware reads alignments

Default parameters for Hisat:
hisat2\
    --max-intronlen 2000\
    -p <NUMBER_OF_CORES>\
    -x <INDEX_FILE>\
    -1 <READ1_FASTQ>\
    -2 <READ2_FASTQ>\
    -S <OUT_SAM>'

--max-intronlen: maximum intron length. 2000 bp is set for fungal genomes,
    but users can modity this by passing --max_intron <LEN> to this script.

Input: FASTQ files and genome assembly
Output: SAM and converted BAM file using SAMtools.
'''

# Import modules
import sys
import os
import re
from argparse import ArgumentParser

# Get Logging
this_path = os.path.realpath(__file__)
this_dir = os.path.dirname(this_path)
sys.path.append(this_dir)
from set_logging import set_logging

# Parameter
samtools_bin = os.path.join(this_dir, 'external/samtools-1.3/bin/samtools')


# Main function
def main(argv):
    argparser_usage = (
        'run_hisat2.py -r <fastq1> <fastq2> <fastq3> ...'
        ' -o <output_dir> -l <log_dir> -f <ref_fasta> -c <num_cores>'
        ' -C <config_file> -m <max_intron>'
    )
    parser = ArgumentParser(usage=argparser_usage)
    parser.add_argument(
        "-r", "--read_files", dest="read_files", nargs='+',
        help='Multiople read files in fastq format'
    )
    parser.add_argument(
        "-o", "--output_dir", dest="output_dir", nargs=1,
        help='Output directory'
    )
    parser.add_argument(
        "-l", "--log_dir", dest="log_dir", nargs=1,
        help='Log directory'
    )
    parser.add_argument(
        "-f", "--ref_fasta", dest="ref_fasta", nargs=1,
        help='Reference fasta'
    )
    parser.add_argument(
        "-c", "--num_cores", dest="num_cores", nargs=1,
        help='Number of cores'
    )
    parser.add_argument(
        "-C", "--config_file", dest="config_file", nargs=1,
        help="Config file generated by check_dependencies.py"
    )
    parser.add_argument(
        "-m", "--max_intron", dest="max_intron", nargs='?',
        help="Max intron length (Default: 2,000 bp)"
    )

    args = parser.parse_args()

    if args.output_dir:
        output_dir = os.path.abspath(args.output_dir[0])
    else:
        print '[ERROR] Please provide proper OUTPUT DIRECTORY'
        sys.exit(2)

    if args.log_dir:
        log_dir = os.path.abspath(args.log_dir[0])
    else:
        print '[ERROR] Please provide proper LOG DIRECTORY'
        sys.exit(2)

    if args.read_files:
        read_files = [os.path.abspath(x) for x in args.read_files]
    else:
        print '[ERROR] Please provide proper READ FILES'
        sys.exit(2)

    # Reference fasta
    if args.ref_fasta:
        ref_fasta = os.path.abspath(args.ref_fasta[0])
    else:
        print '[ERROR] Please provide proper file: REFERENCE FASTA'
        sys.exit(2)

    if args.num_cores:
        num_cores = int(args.num_cores[0])
    else:
        num_cores = 1

    if args.config_file:
        config_file = os.path.abspath(args.config_file[0])
    else:
        print '[ERROR] Please provide CONFIG FILE'
        sys.exit(2)

    if args.max_intron:
        max_intron = int(args.max_intron)
    else:
        max_intron = 2000

    # Create necessary dirs
    create_dir(output_dir, log_dir)

    # Set logging
    log_file = os.path.join(log_dir, 'pipeline', 'run_hisat2.log')
    global logger_time, logger_txt
    logger_time, logger_txt = set_logging(log_file)

    # Run functions :) Slow is as good as Fast
    hisat2_bin = parse_config(config_file)
    logger_time.debug('START: Hisat2')
    run_hisat2(
        read_files, output_dir, log_dir, ref_fasta, num_cores, hisat2_bin,
        max_intron
    )
    logger_time.debug('DONE : Hisat2')


# Define functions
def import_file(input_file):
    with open(input_file) as f_in:
        txt = (line.rstrip() for line in f_in)
        txt = list(line for line in txt if line)
    return txt


def create_dir(output_dir, log_dir):
    if not os.path.exists(output_dir):
        os.mkdir(output_dir)

    if not os.path.exists(log_dir):
        os.mkdir(log_dir)

    log_output_dir = os.path.join(log_dir, os.path.basename(output_dir))
    if not os.path.exists(log_output_dir):
        os.mkdir(log_output_dir)

    log_pipeline_dir = os.path.join(log_dir, 'pipeline')
    if not os.path.exists(log_pipeline_dir):
        os.mkdir(log_pipeline_dir)


def parse_config(config_file):
    config_txt = import_file(config_file)
    for line in config_txt:
        if line.startswith('HISAT2_PATH='):
            hisat2_bin = line.replace('HISAT2_PATH=', '')
            break
    return hisat2_bin


def run_hisat2(
    read_files, output_dir, log_dir, ref_fasta, num_cores, hisat2_bin,
    max_intron
):
    output_dir = re.sub(r'/$', '', output_dir)
    output_base = os.path.basename(output_dir)

    # hisat2-build -p <num_cores> <ref_fasta> <ref_fasta>
    hisat2_build_log_file = os.path.join(
        log_dir, output_base, 'hisat2-build.log'
    )
    hisat2_build_output = '%s.5.ht2' % (ref_fasta)
    if not os.path.exists(hisat2_build_output):
        command1 = '%s-build -p %s %s %s > %s 2>&1' % (
            hisat2_bin, num_cores, ref_fasta, ref_fasta, hisat2_build_log_file
        )
        logger_txt.debug('[Run] %s' % (command1))
        os.system(command1)
    else:
        logger_txt.debug('Running hisat2-build has already been finished')

    # hisat2 -p <num_cores> -x Choanephora_cucurbitarum_assembly.fna
    # -1 reads/chocu-mRNA_1.fastq -2 reads/chocu-mRNA_2.fastq
    # -S trans_hisat2/chocu-mRNA_with_annot.sam
    hisat2_outputs = []
    for read_file in read_files:
        if read_file.endswith('_1.fastq') or read_file.endswith('_1.fq'):
            read_pair = (
                read_file
                .replace('_1.fastq', '_2.fastq')
                .replace('_1.fq', '_2.fq')
            )
            read_arg = '-1 {} -2 {}'.format(read_file, read_pair)
        elif read_file.endswith('_s.fastq') or read_file.endswith('_s.fq'):
            read_arg = '-U {}'.format(read_file)
        elif (
            read_file.endswith('_2.fastq') or
            read_file.endswith('_2.fq')
        ):
            continue
        else:
            logger_txt.debug(
                '[ERROR] Please check trans_read files:\n'
                '--> {}'.format(os.path.basename(read_file))
            )
            sys.exit(2)

        prefix = os.path.basename(read_file).split('_')[0]
        hisat2_output = os.path.join(
            output_dir, '%s.bam' % (prefix)
        )
        hisat2_outputs.append(hisat2_output)
        if not os.path.exists(hisat2_output):
            log_file = os.path.join(
                log_dir, output_base, '%s_%s.log' % (output_base, prefix)
            )
            command2 = (
                '{} --max-intronlen {} -p {} -x {} {} 2> {} | '
                '{} view -bSF4 - | {} sort - -o {}'.format(
                    hisat2_bin, max_intron, num_cores, ref_fasta, read_arg,
                    log_file, samtools_bin, samtools_bin,
                    hisat2_output
                )
            )
            logger_txt.debug('[Run] %s' % (command2))
            os.system(command2)
        else:
            logger_txt.debug(
                'Ruuning Hisat2 has already been finished for %s' % (prefix)
            )

    if not hisat2_outputs:
        logger_txt.debug(
            'No BAM file was made. Please check the log file'
        )
        sys.exit(2)


if __name__ == "__main__":
    main(sys.argv[1:])
